vm-virtio/src/console.rs

```diff
@@ -0,0 +1,353 @@
+// Copyright 2017 The Chromium OS Authors. All rights reserved.
+// Use of this source code is governed by a BSD-style license that can be
+// found in the LICENSE file.
+
+use epoll;
+use libc::EFD_NONBLOCK;
+use std;
+use std::cmp;
+use std::collections::VecDeque;
+use std::io;
+use std::os::unix::io::AsRawFd;
+use std::result;
+use std::sync::atomic::{AtomicUsize, Ordering};
+use std::sync::{Arc, Mutex};
+use std::thread;
+
+use super::Error as DeviceError;
+use super::{
+    ActivateError, ActivateResult, DeviceEventT, Queue, VirtioDevice, VirtioDeviceType,
+    INTERRUPT_STATUS_USED_RING, VIRTIO_F_VERSION_1,
+};
+use crate::VirtioInterrupt;
+use vm_memory::{Bytes, GuestMemoryMmap};
+use vmm_sys_util::EventFd;
+
+const QUEUE_SIZE: u16 = 256;
+const NUM_QUEUES: usize = 2;
+const QUEUE_SIZES: &[u16] = &[QUEUE_SIZE];
+
+// New descriptors are pending on the virtio queue.
+const INPUT_QUEUE_EVENT: DeviceEventT = 0;
+const OUTPUT_QUEUE_EVENT: DeviceEventT = 1;
+// The device has been dropped.
+const KILL_EVENT: DeviceEventT = 2;
+
+struct ConsoleEpollHandler {
+    queues: Vec<Queue>,
+    mem: GuestMemoryMmap,
+    interrupt_status: Arc<AtomicUsize>,
+    interrupt_cb: Arc<VirtioInterrupt>,
+    in_buffer: Arc<Mutex<VecDeque<u8>>>,
+    out: Box<io::Write + Send>,
+    input_queue_evt: EventFd,
+    output_queue_evt: EventFd,
+    kill_evt: EventFd,
+}
+
+impl ConsoleEpollHandler {
+    fn process_input_queue(&mut self) -> bool {
+        let mut in_buffer = self.in_buffer.lock().unwrap();
+        let count = in_buffer.len();
+        let recv_queue = &mut self.queues[0]; //receiveq
```

Comment:
A more verbose comment would be nice here. Just to explain the purpose of the `receiveq` queue.

Reply:
Comment added according to this suggestion.